staging/src/kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1/types.go

```diff
@@ -91,8 +91,17 @@ type StorageSpec struct {
 	// +optional
 	VolumeMode *corev1.PersistentVolumeMode `json:"volumeMode,omitempty"`
 	// This field can be used to specify either: * An existing VolumeSnapshot object (snapshot.storage.k8s.io/VolumeSnapshot) * An existing PVC (PersistentVolumeClaim) * An existing custom resource that implements data population (Alpha) In order to use custom resource types that implement data population, the AnyVolumeDataSource feature gate must be enabled. If the provisioner or an external controller can support the specified data source, it will create a new volume based on the contents of the specified data source.
+	// If the AnyVolumeDataSource feature gate is enabled, this field will always have the same contents as the DataSourceRef field.
 	// +optional
 	DataSource *corev1.TypedLocalObjectReference `json:"dataSource,omitempty"`
+	// Specifies the object from which to populate the volume with data, if a non-empty volume is desired. This may be any local object from a non-empty API group (non core object) or a PersistentVolumeClaim object. When this field is specified, volume binding will only succeed if the type of the specified object matches some installed volume populator or dynamic provisioner.
+	// This field will replace the functionality of the DataSource field and as such if both fields are non-empty, they must have the same value. For backwards compatibility, both fields (DataSource and DataSourceRef) will be set to the same value automatically if one of them is empty and the other is non-empty.
+	// There are two important differences between DataSource and DataSourceRef:
+	// * While DataSource only allows two specific types of objects, DataSourceRef allows any non-core object, as well as PersistentVolumeClaim objects.
+	// * While DataSource ignores disallowed values (dropping them), DataSourceRef preserves all values, and generates an error if a disallowed value is specified.
+	// (Beta) Using this field requires the AnyVolumeDataSource feature gate to be enabled.
+	// +optional
+	DataSourceRef *corev1.TypedLocalObjectReference `json:"dataSourceRef,omitempty"`
```

Comment:
I'm not saying this is wrong, but it feels weird to me that the datavolume `spec.source` will be unset when using this API.  My assumption was that `DataVolumeSource` would have a `Populator` field.

Reply:
I also like the idea of having a `Populator` field but in practice I found it to be redundant. I have another semi-finished implementation where I add the new source so changing it wouldn't really be a problem, but I found some drawbacks: 

- Since we are embedding the PVC spec into our Storage API, it makes sense to also embed the `DataSourceRef` field (it's already supported in the PVC API). We should then allow users to populate it, which would leave the source field useless.
- By having a `Populator` source, we could allow users to specify the populator there in the same way as they'd do it on the `DataSourceRef` field, and then copy it to the PVC spec. However, this would require additional validation in case both the `DataSourceRef` and `source.Populator` fields are set without really providing any advantage. 
- The biggest drawback for me: We would basically allow to do the same thing in two different ways without any justification or real advantages, which in my opinion is an ugly practice.

I still see some advantages: Having a `Populator` source is easier to understand for users, and in practice, they'd probably be using that field instead of `DataSourceRef`. Eventually, when we adopt populators as the main population method in CDI, we could just use that single `Populator` source for all cases, wich in my opinion is prettier than `DataSourceRef`. Let me know what you think.